app/integrations.go

```diff
@@ -0,0 +1,170 @@
+package app
+
+import (
+	"encoding/json"
+	"net/http"
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/mattermost/mattermost-server/v6/app/request"
+	"github.com/mattermost/mattermost-server/v6/model"
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+type ListedApp struct {
+	Manifest struct {
+		AppID       string `json:"app_id"`
+		DisplayName string `json:"display_name"`
+		Version     string `json:"version"`
+	}
+
+	Installed bool `json:"installed"`
+	Enabled   bool `json:"enabled"`
+}
+
+func (a *App) GetInstalledIntegrations() ([]*model.InstalledIntegration, *model.AppError) {
+	return a.ch.getInstalledIntegrations()
+}
+
+func (ch *Channels) getInstalledIntegrations() ([]*model.InstalledIntegration, *model.AppError) {
+	out := []*model.InstalledIntegration{}
+
+	pluginsEnvironment := ch.GetPluginsEnvironment()
+	if pluginsEnvironment == nil {
+		return out, nil
+	}
+
+	plugins, err := pluginsEnvironment.Available()
+	if err != nil {
+		return nil, model.NewAppError("getInstalledIntegrations", "app.plugin.sync.read_local_folder.app_error", nil, err.Error(), 0)
+	}
+
+	for _, p := range plugins {
+		ignore := false
```

Comment:
@neallred Although the nested loop is a little ugly, that's how most of the codebase is written, and is common with Go code AFAIK. 0/5 on whether we should have it either way. Do we have a set pattern for this in our codebase? (No pun intended)

Reply:
I see what you did there :smile: The set pattern is also common in go, and we use it in a handful of places in mattermost-server:

```
app/channel.go
3185:	channelMemberMap := make(map[string]struct{})

app/product.go
32:	pmap := make(map[string]struct{})

store/sqlstore/channel_store.go
1458:		visited := make(map[string]struct{})

services/sharedchannel/sync_send.go
79:	notified := make(map[string]struct{})

model/user.go
853:var restrictedUsernames = map[string]struct{}{
```

The empty struct doesn't take space, so this set structure should be comparable in size to a slice.

Its good to avoid nested loops for brevity, readability, and because they are often quadratic time. For example, every time we add a new system plugin, it is likely to exist in the plugins we fetch, and will always exist in the list/set of system plugins. So if we have 5 system plugins, that's up to 25 iterations between the two loops. If we increase to 6, that's up to 36. If we have the set, it becomes linear, 5 -> 6. In this case, the number of plugins is small enough that it shouldn't matter much for performance either way.